processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -52,333 +56,313 @@
-          }
-          continue;
-        }
-        if (lookupSnapshotTaker != null) {
-          lookupSnapshotTaker.takeSnapshot(getAllAsList());
-        }
+      if (!queue.offer(new LoadNotice(lookupName, lookupExtractorFactoryContainer), 1, TimeUnit.MILLISECONDS)) {
+        throw new ISE("notice queue add timedout to add [%s] lookup drop notice", lookupName);
       }
-      if (!failedExtractorFactoryMap.isEmpty()) {
-        throw new ISE(
-            "was not able to start the following lookup(s) [%s]",
-            failedExtractorFactoryMap.keySet().toString()
-        );
-      }
-    }
-    finally {
-      startStopLock.readLock().unlock();
+    } catch (InterruptedException ex) {
+      throw new ISE(ex, "failed to add [%s] lookup load notice", lookupName);
     }
   }
 
-  /**
-   * Add or update a lookup factory
-   *
-   * @param lookupName             The name of the lookup
-   * @param lookupExtractorFactory The factory of the lookup
-   *
-   * @return True if the lookup was updated, false otherwise
-   *
-   * @throws IllegalStateException if start of the factory fails
-   */
-  public boolean updateIfNew(String lookupName, final LookupExtractorFactory lookupExtractorFactory)
+  public void remove(String lookupName)
   {
-    boolean update = false;
+    Preconditions.checkState(lifecycleLock.isStarted());
+
     try {
-      startStopLock.readLock().lockInterruptibly();
-    }
-    catch (InterruptedException e) {
-      throw Throwables.propagate(e);
+      if (!queue.offer(new DropNotice(lookupName), 1, TimeUnit.MILLISECONDS)) {
+        throw new ISE("notice queue add timedout to add [%s] lookup drop notice", lookupName);
+      }
+    } catch (InterruptedException ex) {
+      throw new ISE(ex, "failed to add [%s] lookup drop notice", lookupName);
     }
-    try {
-      assertStarted();
-      LookupExtractorFactory prior = lookupMap.get(lookupName);
-      update = lookupExtractorFactory.replaces(prior);
-      if (update) {
-        if (!lookupExtractorFactory.start()) {
-          throw new ISE("Could not start [%s]", lookupName);
-        }
-        boolean racy;
-        do {
-          if (prior == null) {
-            racy = null != lookupMap.putIfAbsent(lookupName, lookupExtractorFactory);
-          } else {
-            racy = !lookupMap.replace(lookupName, prior, lookupExtractorFactory);
-          }
+  }
 
-          if (racy) {
-            prior = lookupMap.get(lookupName);
-            update = lookupExtractorFactory.replaces(prior);
-          }
-        } while (racy && update);
+  @Nullable
+  public LookupExtractorFactoryContainer get(String lookupName)
+  {
+    Preconditions.checkState(lifecycleLock.isStarted());
+    return lookupMap.get(lookupName);
+  }
 
-        if (prior != null && update) {
-          if (!prior.close()) {
-            LOGGER.error("Error closing [%s]:[%s]", lookupName, prior);
-          }
-        }
+  public LookupsState getAllLookupsState()
```

Comment:
Now without locks this method may return duplicate entries in the returned collections, because some queue elements are processed by mainThread while iteration is in progress.

For more consistency, unique noticeId could be added to notices, it is obtained from some AtomicLong via getAndIncrement(), and then after obtaining local copies of collections, entries from `lookupsToLoad` and `lookupsToDrop` should be filtered, the criteria is that the corresponding entry in the lookupMap has higher noticeId.